src/Avalonia.Animation/AnimatorStateMachine`1.cs

```diff
@@ -1,273 +1,208 @@
+                    repeatCount = (long)animation.RepeatCount.Value;
                     break;
             }
 
-            _isReversed = (animation.PlaybackDirection & PlaybackDirection.Reverse) != 0;
-            _animationDirection = _targetAnimation.PlaybackDirection;
-            _fillMode = _targetAnimation.FillMode;
-
-            if (_durationTotalFrameCount > 0)
-                _currentState = KeyFramesStates.DoDelay;
-            else
-                _currentState = KeyFramesStates.DoRun;
-            _onComplete = onComplete;
+            animationDirection = animation.PlaybackDirection;
+            fillMode = animation.FillMode; 
+            this.onComplete = onComplete;
         }
 
-        public void Step(PlayState _playState, Func<double, T, T> Interpolator)
+        public void Step(PlayState playState, long frameTick, Func<double, T, T> Interpolator)
         {
             try
             {
-                InternalStep(_playState, Interpolator);
+                InternalStep(playState, frameTick, Interpolator);
             }
             catch (Exception e)
             {
-                _targetObserver?.OnError(e);
+                targetObserver?.OnError(e);
             }
         }
 
-        private void InternalStep(PlayState _playState, Func<double, T, T> Interpolator)
+        private void DoComplete()
         {
-            if (!_gotFirstKFValue)
+            if (fillMode == FillMode.Forward || fillMode == FillMode.Both)
+                targetControl.SetValue(parent.Property, lastInterpValue, BindingPriority.LocalValue);
+
+            targetObserver.OnCompleted();
+            onComplete?.Invoke();
+            Dispose();
+        }
+
+        private void DoDelay()
+        {
+            if (fillMode == FillMode.Backward || fillMode == FillMode.Both)
+                if (currentIteration == 0)
+                    targetObserver.OnNext(firstKFValue);
+                else
+                    targetObserver.OnNext(lastInterpValue);
+        }
+
+        private void InternalStep(PlayState playState, long frameTick, Func<double, T, T> Interpolator)
+        {
+            if (!gotFirstKFValue)
+            {
+                firstKFValue = (T)parent.First().Value;
+                gotFirstKFValue = true;
+            }
+
+            if (!gotFirstFrameCount)
             {
-                _firstKFValue = _parent.First().Value;
-                _gotFirstKFValue = true;
+                firstFrameCount = frameTick;
+                gotFirstFrameCount = true;
             }
 
-            if (_currentState == KeyFramesStates.Disposed)
+            if (isDisposed)
                 throw new InvalidProgramException("This KeyFrames Animation is already disposed.");
 
-            if (_playState == PlayState.Stop)
-                _currentState = KeyFramesStates.Stop;
+            if (playState == PlayState.Stop)
+                DoComplete();
 
             // Save state and pause the machine
```

Comment:
I'd prefer not to have commented-out code in master as it quickly becomes stale, and it's not obvious _why_ it was commented out.

Reply:
Sorry about that sir, still in progress and forgot to place a TODO. This will not be in the final PR